netsim/extra/node.clone/plugin.py

```diff
@@ -0,0 +1,109 @@
+from netsim import data
+from netsim.utils import log,strings
+from netsim.augment import links
+
+def clone_interfaces(link_data: Box, nodename: str, clones: list[str]) -> list[Box]:
+  ifs = []
+  if nodename in [ i.node for i in link_data.get('interfaces',[]) ]:
+    for c,clone in enumerate(clones):
+      l = data.get_box(link_data)
+      l.interfaces = []
+      for intf in link_data.interfaces:
+        intf_clone = data.get_box(intf)
+        if intf.node == nodename:
+          intf_clone.node = clone
+        elif 'ifindex' in intf:
+          intf_clone.ifindex = intf.ifindex + c
+        l.interfaces.append(intf_clone)
+      ifs.append(l)
+  return ifs
+
+"""
+update_links - updates 'links' lists in VLAN and VRF objects
+"""
+def update_links(topo_items: Box, nodename: str, clones: list, topology: Box) -> None:
+  for vname,vdata in topo_items.items():                                    # Iterate over global VLANs or VRFs
+    if not isinstance(vdata,Box):                                           # VLAN not yet a dictionary?
+      continue                                                              # ... no problem, skip it
+    if not 'links' in vdata:                                                # No VLAN links?
+      continue                                                              # ... no problem, move on
+
+    for cnt,l in enumerate(list(vdata.links)):                              # So far so good, now iterate over the links
+      link_data = links.adjust_link_object(                                 # Create link data from link definition
+                    l=l,
+                    linkname=f'vlans.{vname}.links[{cnt+1}]',
+                    nodes=topology.nodes)
+      if link_data is not None:
+        ifs = clone_interfaces(link_data,nodename,clones)
+        if ifs:
+          vdata.links.remove(l)
+          vdata.links += ifs
+
+"""
+clone_node - Clones a given node N times, creating additional links and/or interfaces for the new nodes
+"""
+def clone_node(node: Box, topology: Box) -> None:
+  _p = { 'start': 1, 'step': 1 } + node.pop('clone',{})                      # Define cloning parameters
+  if 'count' not in _p:
+    log.error("Node {node.name} missing required attribute clone.count",     # Not validated by Netlab yet
+              category=AttributeError, module='node.clone')
+    return
+
+  if 'include' in node:                                                      # Check for components
+    log.error("Cannot clone component {node.name}, only elementary nodes",
+              category=AttributeError, module='node.clone')
+    return
+
+  name_format = topology.defaults.clone.node_name_pattern
+
+  orig_name = node.name
+  node.name = strings.eval_format(name_format, node + { 'id': _p.start } )  # Rename first node as clone_01
+  topology.nodes += { node.name: node }
+
+  clones = [ node.name ]
+  for c in range(_p.start+_p.step,_p.start+_p.count*_p.step,_p.step):       # Existing node is '1'
```

Comment:
I don't see why you treat the first node differently than the others? At the very minimum, it would be nice to have a comment to explain it.

Reply:
Leftover from earlier embodiment, it should be removed